core/metacling/src/TCling.cxx

```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
-      }
+   // Preloading the system modules will give us similar behavior to having a PCH.
+   headerSearch.loadTopLevelSystemModules();
 
-      Error("TCling::LoadCoreModules",
-            "Internal error, couldn't find core "
-            "C++ modules in modulemap:%s\n",
-            MissingModuleNameList.c_str());
-   }
-
-   // Collect all submodules in the found core modules.
-   for (size_t i = 0; i < coreModules.size(); ++i) {
-      for (clang::Module *subModule : coreModules[i]->submodules())
-         coreModules.push_back(subModule);
-   }
-
-   // Now we collect all header files from the previously collected modules.
-   std::vector<StringRef> moduleHeaders;
-   for (clang::Module *module : coreModules) {
-      ROOT::TMetaUtils::foreachHeaderInModule(
-         *module, [&moduleHeaders](const clang::Module::Header &h) { moduleHeaders.push_back(h.NameAsWritten); });
-   }
-
-   // Turn the list of found header files into C++ code that includes all of
-   // them. This will be wrapped into an `import` declaration by clang, so we
-   // only make those modules available, not actually textually include those
-   // headers.
-   // FIXME: Use clang::ASTReader::makeModuleVisible.
-   std::stringstream declarations;
-   for (StringRef H : moduleHeaders) {
-      declarations << "#include \"" << H.str() << "\"\n";
-   }
-
-   // C99 decided that it's a very good idea to name a macro `I` (the letter I).
-   // This seems to screw up nearly all the template code out there as `I` is
-   // common template parameter name and iterator variable name.
-   // Let's follow the GCC recommendation and undefine `I` in case any of the
-   // core modules have defined it:
-   // https://www.gnu.org/software/libc/manual/html_node/Complex-Numbers.html
-   declarations << "#ifdef I\n #undef I\n #endif\n";
-
-   auto result = interp.declare(declarations.str());
+   cling::Transaction* T = nullptr;
+   interp.declare("/*This is decl is to get a valid sloc...*/;", &T);
+   SourceLocation ValidLoc = T->decls_begin()->m_DGR.getSingleDecl()->getLocStart();
+   clang::Sema &TheSema = CI.getSema();
+   // createImplicitModuleImportForErrorRecovery creates decls.
+   cling::Interpreter::PushTransactionRAII RAII(&interp);
+   std::vector<clang::Module*> TopLevelModulesToLoad;
+   for (auto I = moduleMap.module_begin(), E = moduleMap.module_end(); I != E; ++I) {
+     clang::Module *M = I->second;
+     assert(M);
+     if (M->IsSystem && !M->IsMissingRequirement) {
+       TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, M);
+       for (clang::Module *subModule : M->submodules())
+         TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, subModule);
+     }
+     else if (M->IsSystem && !M->IsMissingRequirement)
+       TopLevelModulesToLoad.push_back(M);
+   }
+   // Those two modules are required to be preloaded by ROOT. If they are not around we have a big problem.
+   clang::Module *CoreM = moduleMap.findModule("Core");
```

Comment:
Why not recursively loading the submodules here? Does it work? (never tested that).